src/main/scala/wdl4s/WdlNamespace.scala

```diff
@@ -197,85 +167,297 @@ object WdlNamespace {
+
+    all.toSeq.sortWith({ case (l, r) => l.getMessage < r.getMessage}) match {
+      case s: Seq[SyntaxError] if s.nonEmpty => throw s.head
+      case _ =>
+    }
+
+    namespace
+  }
+
+  private def getDecls(scope: Scope): Seq[DeclarationInterface] = {
+    scope match {
+      case t: Task => t.declarations ++ t.outputs
+      case s => s.declarations
+    }
+  }
+
+  private def declarationName(declarationAst: Ast): Terminal = declarationAst.getAttribute("name").asInstanceOf[Terminal]
+
+  def validateDeclaration(declaration: DeclarationInterface, wdlSyntaxErrorFormatter: WdlSyntaxErrorFormatter): Seq[SyntaxError] = {
+    val invalidVariableReferences = for {
+      expr <- declaration.expression.toSeq
+      variable <- expr.variableReferences
+      if declaration.resolveVariable(variable.sourceString).isEmpty
+    } yield new SyntaxError(wdlSyntaxErrorFormatter.declarationContainsInvalidVariableReference(
+      declarationName(declaration.ast),
+      variable
+    ))
+
+    val typeMismatches = typeCheckDeclaration(declaration, wdlSyntaxErrorFormatter).toSeq
+
+    invalidVariableReferences ++ typeMismatches
+  }
+
+  def typeCheckDeclaration(decl: DeclarationInterface, wdlSyntaxErrorFormatter: WdlSyntaxErrorFormatter): Option[SyntaxError] = {
+    def lookupType(from: Scope)(n: String): WdlType = {
+      from.resolveVariable(n) match {
+        case Some(d: DeclarationInterface) => d.wdlType
+        case Some(c: Call) => WdlObjectType
+        case Some(s: Scatter) => s.collection.evaluateType(lookupType(s), new WdlStandardLibraryFunctionsType) match {
+          case Success(a: WdlArrayType) => a.memberType
+          case _ => throw new VariableLookupException(s"Variable $n references a scatter block ${s.fullyQualifiedName}, but the collection does not evaluate to an array")
+        }
+        case Some(ns: WdlNamespace) => WdlNamespaceType
+        case _ => throw new VariableLookupException(s"Could not resolve $n from scope ${from.fullyQualifiedName}")
+      }
     }
 
-    // FIXME: Here's where I'd toSet stuff after duplications are detected
-    ast.findAsts(AstNodeName.Workflow) match {
-      case Nil => NamespaceWithoutWorkflow(namespace, imports, namespaces, tasks, terminalMap, ast)
-      case Seq(x) => NamespaceWithWorkflow(ast, x, namespace, imports, namespaces, tasks, terminalMap, wdlSyntaxErrorFormatter)
-      case doh => throw new SyntaxError(wdlSyntaxErrorFormatter.tooManyWorkflows(doh.asJava))
+    decl.expression flatMap { expr =>
+      expr.evaluateType(lookupType(decl), new WdlStandardLibraryFunctionsType) match {
+        case Success(wdlType) if !decl.wdlType.isCoerceableFrom(wdlType) =>
+          Option(new SyntaxError(wdlSyntaxErrorFormatter.taskOutputExpressionTypeDoesNotMatchDeclaredType(
+            declarationName(decl.ast), wdlType, decl.wdlType
+          )))
+        case Success(wdlType) =>
+          expr.evaluate((s: String) => throw new Throwable("not implemented"), NoFunctions) match {
```

Comment:
Yeah I'll replace with `NoLookup`. This is a "best effort" kind of validation, in the sense that we try to evaluate the declaration without lookup nor wdl function, if it succeeds we check that the result type can be coerced to the target type and if not we move on.